### PR TITLE
fix Issue 20963 - wrong code for cast(double)anUlong

### PR DESCRIPTION
**Pull request by WalterBright**

### CODE REVIEW

File: src/dmd/backend/cgelem.d [hunk 1]
```diff
@@ -4790,34 +4790,66 @@ private elem * el32_64(elem *e, goal_t goal)
 
 private elem *elu64_d(elem *e, goal_t goal)
 {
-    if (e.EV.E1.Eoper != OPconst && (I64 || (I32 && config.inline8087)))
+    if (e.EV.E1.Eoper != OPconst)
```
Comment (thewilsonator) [inline src/dmd/backend/cgelem.d:4793]: Use an early return.
```
if (e.EV.E1.Eoper != OPconst)
    return evalu8(e, goal);
...
```

**Comment by thewilsonator** [comment]
stable

**Comment by WalterBright** [comment]
> stable

It's not a regression.

**Comment by NilsLankila** [comment]
But it's a serious bug. It's been more than one year that even "normal" bugs are accepted in stable.

**Comment by WalterBright** [comment]
It's been there for 30+ years. How serious can it be?

**Comment by NilsLankila** [comment]
ok, appologies. It's just that I've been said that in the past, I quote "any bug fix can go in stable".
Anyway, let's not ruin your PR with this debate.

**Comment by WalterBright** [comment]
Things are looking more complex than I originally thought. Going to put this on hold for the moment.

File: src/dmd/backend/cgelem.d [hunk 2]
```diff
@@ -4790,34 +4790,66 @@ private elem * el32_64(elem *e, goal_t goal)
...
+            u2 = el_bin(OPor, TYllong, u2, u3);
+            u2 = el_una(OPs64_d, e.Ety, u2);
+            u2 = el_bin(OPmul, e.Ety, u2, el_long(e.Ety, 2));
+
+            e.Eoper = OPcond;
+            e.EV.E1 = u;
+            e.EV.E2 = el_bin(OPcolon, e.Ety, u1, u2);
+            return optelem(e, GOALvalue);
+        }
+        else if (config.inline8087)
```
Comment (UplinkCoder) [inline src/dmd/backend/cgelem.d:4823]: the else is not required it seems.

File: src/dmd/backend/cgelem.d [hunk 3]
```diff
@@ -4804,20 +4830,64 @@ private elem *elu64_d(elem *e, goal_t goal)
...
+        u2 = el_bin(OPmul, ty, u2, el_long(ty, 2));
+
+        elem* r = el_bin(OPcond, e.Ety, u, el_bin(OPcolon, e.Ety, u1, u2));
+        *pu = null;
+        el_free(e);
+        return optelem(r, GOALvalue);
     }
-    else
-        return evalu8(e, goal);
+    else if (config.inline8087)
```
Comment (andralex) [inline src/dmd/backend/cgelem.d:4853]: yah this is redundant, should be removed

```suggestion
    if (config.inline8087)
```

File: src/dmd/backend/cgelem.d [hunk 4]
```diff
@@ -4804,20 +4830,64 @@ private elem *elu64_d(elem *e, goal_t goal)
...
+        if (ty != TYldouble)
+            r = el_una(OPtoprec, e.Ety, r);
+
+        *pu = null;
+        el_free(e);
+
+        return optelem(r, GOALvalue);
+    }
+
+    return evalu8(e, goal);
```
Comment (andralex) [inline src/dmd/backend/cgelem.d:4890]: ...especially because there's no `else` here.

File: src/dmd/backend/evalu8.d [hunk 5]
```diff
@@ -425,6 +425,7 @@ static if (0)
                         e.EV.Vcldouble.im = 0  + e2.EV.Vcldouble.im;
                         break;
                     default:
+elem_print(e);
```
Comment (UplinkCoder) [inline src/dmd/backend/evalu8.d:428]: this looks like an oversight.

Comment (andralex) [reply src/dmd/backend/evalu8.d:428]:
```suggestion
```

**Review by andralex — APPROVED** [review]
I took the liberty to remove the elem_print line, thanks @UplinkCoder

File: test/runnable/testfloat.d [hunk 6]
```diff
@@ -207,6 +207,17 @@ void test20349()
...
 
+/****************************************/
+// https://issues.dlang.org/show_bug.cgi?id=20963
+
+void test20963()
+{
+    ulong v = 0xE3251BACB112CB8B;
+    double d = cast(double)v;
+    printf("%a\n", d); //0x1.c64a37596225ap+63
+    assert(d == 0x1.c64a375962259p+63);
```
Comment (UplinkCoder) [inline test/runnable/testfloat.d:218]: this test seems oddly implementation specific.

**Pull request merged**